### DISCUSSION TITLE
Calculator on each selection

### DISCUSSION THREAD

**Discussion by carrascomj** [helix-editor/helix#6879] [Ideas] [answered]
## Use case

When manipulating CSVs or tables in latex, markdown, etc. sometimes I want to compute a numerical operation on columns.

## Feature

Evaluate an arithmetic expression on each selection on the typable command "calc" (similar to "pipe").

![t-rec](https://user-images.githubusercontent.com/46683255/234308140-6e0a377e-5490-4fe8-a229-dcfacb6f65b5.gif)

####  Design details

* In the prompt, "%" represents each range fragment. This should definitely be changed because it is an actual operator (and it has a very different and vim which one may want to avoid).
* After replacement, the expression gets evaluated using the [`eval`](https://docs.rs/eval/latest/eval/) crate, which introduces another dependency.
* In the example above I am replacing the ranges, but I would prefer to copy them to the default register.

I have grepped through kakoune's documentation at Gtihub without any luck so I don't know about the prior art. If this feature gets fleshed out and there's interest I can submit a PR.

This behaviour can be achieved using the `:pipe` command and [`bc`](https://man7.org/linux/man-pages/man1/bc.1p.html) (maybe with `xargs`), but I have never ever got it right.

**Comment by pascalkuthe** [comment, marked as answer]
Something like this should wait for the built-in scripting language which would be used instead of eval here.

**Reply by carrascomj** [reply]
That's probably the best option since the eval crate has been abandoned for some time and the feature is a bit niche.

It is still very handy. Something like this plus virtual space for indentation and some syntax highligting would make a great `csv.hx` plugin.

**Comment by clouedoc** [comment]
This is a really cool feature I was looking for today for some accounting. I now it's been forever, but do you still have the branch where you implemented it @carrascomj ? I'd like to port it to the current version for my personal use and perhaps opening a PR :)

**Reply by carrascomj** [reply]
I removed it sorry. I think it was easy to implement (especially since I used the `eval` crate). Probably, you would want to create a plugin in steel: https://github.com/helix-editor/helix/pull/8675

Edit: you can also put in the path an executable bash script called `calc` containing the following:

```bash
#!/usr/bin/env bash

# Usage: echo 5 | ./calc.sh 'SEL * 3.7 + 2'

val=$(cat)                # read single value from stdin (pipe)
expr_template=$1          # expression containing SEL
expr_expanded=${expr_template//SEL/$val}

printf 'scale=3\n%s\n' "$expr_expanded" | bc -l
```

And then you can `:pipe` selections inside helix with an argument like `pipe: calc "SEL + 2"`

**Reply by clouedoc** [reply]
Cool, thanks!

**Comment by Rudxain** [comment]
About prior art, VSCode has had a feature to evaluate expressions for some time. I can't find the docs, but it seems it's powered by Emmet
